cpp/src/shap/kernel_shap.cu

```diff
@@ -0,0 +1,259 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cuml/explainer/kernel_shap.hpp>
+
+#include <curand.h>
+#include <curand_kernel.h>
+
+namespace ML {
+namespace Explainer {
+
+/*
+* Kernel distrubutes exact part of the kernel shap dataset
+* Each block scatters the data of a row of `observations` into the (number of rows of
+* background) in `dataset`, based on the row of `X`.
+* So, given:
+* background = [[0, 1, 2],
+                [3, 4, 5]]
+* observation = [100, 101, 102]
+* X = [[1, 0, 1],
+*      [0, 1, 1]]
+*
+* dataset (output):
+* [[100, 1, 102],
+*  [100, 4, 102]
+*  [0, 101, 102],
+*  [3, 101, 102]]
+*
+*
+*/
+template <typename DataT, typename IdxT>
+__global__ void exact_rows_kernel_sm(DataT* X, IdxT nrows_X, IdxT ncols,
+                                     DataT* background, IdxT nrows_background,
+                                     DataT* dataset, DataT* observation) {
+  extern __shared__ int idx[];
+  int i, j;
+
+  if (threadIdx.x < nrows_background) {
+    // the first thread of each block gets the row of X that the block will use
+    // for the scatter.
+    if (threadIdx.x == 0) {
+      for (i = 0; i < ncols; i++) {
+        idx[i] = (int)X[blockIdx.x * ncols + i];
+      }
+    }
+    __syncthreads();
+
+    // all the threads now scatter the row, based on background and new observation
+    int row = blockIdx.x * nrows_background + threadIdx.x;
+#pragma unroll
```

Comment:
Does this get unrolled when the number of iterations is not specified with a compile-time constant?

Reply:
no, I left it unspecified from a prior version (incorrectly), it should be partially unrolled